### PR TITLE
Async overhaul

### LINKED ISSUE #2279: Automatic night mode broken?
There have been several reports of this. 

http://www.reddit.com/r/Enhancement/comments/39fccw/how_do_i_customise_night_mode/

### LINKED ISSUE #2291: Asynchronous storage
@honestbleeps and I need to spec out what that actually means..

### LINKED ISSUE #2497: Opening a link (consisting of only a hash) in a new tab opens the background page
https://www.reddit.com/r/RESissues/comments/3qt3e5/bug_selecting_a_markdown_link_with_an_invalid_url/

Basically, because RES has to hit the API from the [background page](https://github.com/honestbleeps/Reddit-Enhancement-Suite/blob/ed66de559a2d3aac86a445221a6dcd0a1ba28dd5/Chrome/background.js#L177) to open new tabs, a link to `#foo` will go to `_generated_background_page.html#foo` when it's opened in a new tab using keyboard shortcuts.

For some reason, `chrome.tabs.create` seems to resolve relative URLs (`/bar`) based on the URL of the tab sending the request, but hashes (`#foo`) based on the background page's URL. ¯\\_(ツ)_/¯

### LINKED ISSUE #2533: Minor: DeviantArt Media Host will silently swallow errors raised in its `handleInfo` method
Minor issue I discovered while working on #2531: if an error is thrown in DeviantArt's handleInfo method, the error will be silently caught and ignored thanks to https://github.com/honestbleeps/Reddit-Enhancement-Suite/blob/master/lib/modules/hosts/deviantart.js#L16-L21

### LINKED ISSUE #2672: Vidble albums link directly to the image instead of the album
https://github.com/honestbleeps/Reddit-Enhancement-Suite/issues/2605#issuecomment-170937482

I'll just throw this in for 4.7.

### PR DESCRIPTION
**Pull request by erikdesjardins**
Alt title: "all hacks must die".

Fixes #2279, fixes #2291, fixes #2497, fixes #2533, fixes #2672, closes #2536, closes #2664.
Depends on #2454, #2507, #2575.
Depends on Babel issues: ~~[T2776](https://phabricator.babeljs.io/T2776) (fixed, needs to be merged and released), [T6882](https://phabricator.babeljs.io/T6882) (I'll look into contributing a fix for this soon).~~
~~Due to these issues, it won't work with the current set of transformers - if you want to build this branch, you'll need to replace your `.babelrc` with [this](https://gist.github.com/erikdesjardins/94843e8aa51e00a427ae), which will only run in browsers with generator support.~~
~~Both fixed and released.~~
~~Not quite. [T3077](https://phabricator.babeljs.io/T3077) / [T6972](https://phabricator.babeljs.io/T6972)~~
Babel 6.6.0 is out, time is go

Contains major rewrites/refactoring for all backends, RESEnvironment, init, and media hosts.

In short: async storage (minor performance improvements), all caching moved into memory, large performance improvements for init, backends are much cleaner, safer, and easier to debug (a *ton* of hacks were removed, all browsers now present an identical messaging API with proper responses).

Many more details in the commit messages.

### CODE REVIEW

**Comment by honestbleeps** [comment]
oh man I've been dreaming of doing this for so long... I haven't reviewed yet, but I'm super excited about this.

This would be a good opportunity (either in this PR or after it is merged) to start using `chrome.storage` instead of background page localStorage, and to move from Firefox's tremendously unstable simple-storage to indexedb, as well.  Those really relied on an async refactor.

**Comment by Ajedi32** [comment]
Woah. :open_mouth: Nice.

**Comment by erikdesjardins** [comment]
Oh yeah - I have no idea if the Safari backend works. I'm gonna try running OS X in VirtualBox tonight to test it.

**Comment by mc10** [comment]
@erikdesjardins It might be better to squash this into fewer but larger commits to simplify commit history, but it's obviously up to you.

**Comment by erikdesjardins** [comment]
@mc10 Yeah, I'll see what I can do about that.

**Comment by erikdesjardins** [comment]
Actually I'd rather keep them separate so the intent of each commit is more obvious.
I care more about having a descriptive blame than a clean history.

**Comment by mc10** [comment]
That's fair. Also, the mentioned babel issues have been resolved, so is babel good to go?

**Comment by Ajedi32** [comment]
Pretty sure the Babel fix needs to be released first right? [4142003bbd2fe3f862f9ddfde05274ee003de2cc](https://github.com/babel/babel/commit/4142003bbd2fe3f862f9ddfde05274ee003de2cc) is just on master right now.

**Comment by Ajedi32** [comment]
Also, are you sure [T2776](https://phabricator.babeljs.io/T2776) is fixed? The issue is still open.

**Comment by erikdesjardins** [comment]
@mc10 You make a good point though; I just think it would be hard to do it right for this PR (cause it's so broad).

@Ajedi32 I guess they forgot to close it (or they're not going to until the release), it's fixed by [this PR](https://github.com/babel/babel/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+author%3Agzzhanghao+generators).

From what I hear, they're planning to release it on Sunday.
